feat(git): add push-remote ahead/behind tracking

sanki92 opened the pull request:
### Prerequisites

- [x] I have read and understood the [contributing guide][CONTRIBUTING.md].
- [x] The commit message follows the [conventional commits][cc] guidelines.
- [x] Tests for the changes have been added (for bug fixes / features).
- [x] Docs have been added/updated (for bug fixes / features).

### Description

## Issue
Closes #5312

## Problem
The Git segment only provided `.Ahead` and `.Behind` values for the fetch-remote. When working with separate fetch and push remotes, users had no way to see how many commits they were ahead or behind of their push-remote.

## Solution
Added support for tracking commits ahead/behind for the push-remote by introducing two new template properties:
- `.PushAhead` - Number of commits ahead of the push-remote
- `.PushBehind` - Number of commits behind the push-remote

The implementation:
- Uses `git rev-list --count` to accurately calculate ahead/behind commits
- Respects Git's push-remote configuration hierarchy:
  - Branch-specific `pushRemote` config
  - Global `remote.pushDefault` config
  - Falls back to fetch-remote if no push-remote is configured
- Only executes when `fetch_status` is enabled
- Uses the same efficient approach as powerlevel10k

## Changes Made
- **Modified `src/segments/git.go`**: 
  - Added `PushAhead` and `PushBehind` fields to Git struct
  - Implemented `setPushStatus()` to calculate push-remote ahead/behind
  - Implemented `getPushRemote()` to detect the correct push-remote branch
  - Integrated push status calculation into the existing status flow
- **Modified `src/segments/git_test.go`**: 
  - Added `TestPushStatusAheadAndBehind` with comprehensive test cases
  - Tests cover: ahead only, behind only, both, and up-to-date scenarios

## Testing
- ✅ All existing Git segment tests pass (no breaking changes)
- ✅ New unit test validates the calculation logic
- ✅ Project builds successfully without errors
- ✅ Tested scenarios: ahead, behind, both, and synchronized states

## Usage Example
Users can now access push-remote status in their templates:
```json
{
  "type": "git",
  "properties": {
    "fetch_status": true
  },
  "template": "{{ .HEAD }} Fetch:↑{{ .Ahead }}↓{{ .Behind }} Push:↑{{ .PushAhead }}↓{{ .PushBehind }}"
}
```

[CONTRIBUTING.md]: https://github.com/JanDeDobbeleer/oh-my-posh/blob/main/CONTRIBUTING.md
[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

sanki92 committed:
feat(git): add push-remote ahead/behind tracking

sanki92 committed:
docs(git): add PushAhead and PushBehind properties

sanki92 committed:
refactor(git): use filepath.Join for config path

sanki92 commented on src/segments/git.go:
```diff
@@ -538,6 +541,59 @@ func (g *Git) setBranchStatus() {
+}
+
+func (g *Git) getPushRemote() string {
+	upstream := regex.ReplaceAllString("/.*", g.Upstream, "")
```
The original code is correct. The `regex.ReplaceAllString` function signature in this codebase is `ReplaceAllString(pattern, text, replaceText)` (defined in `src/regex/regex.go`), so the pattern `/.*` must come first. This correctly strips everything after the slash from `g.Upstream` (e.g., `origin/main` → `origin`). The same pattern is used consistently elsewhere in the code (see line 997 in `getRemoteURL()`).

sanki92 commented on src/segments/git.go:
```diff
@@ -538,6 +541,59 @@ func (g *Git) setBranchStatus() {
+		return ""
+	}
+
+	cfg, err := ini.Load(g.scmDir + "/config")
```
Good catch! Applied this suggestion in commit c49542ad. Using `filepath.Join` is more secure and handles path separators correctly across different operating systems. Thanks for the review!

JanDeDobbeleer commented on src/segments/git.go:
```diff
@@ -538,6 +541,59 @@ func (g *Git) setBranchStatus() {
 	g.BranchStatus = getBranchStatus()
 }
 
+func (g *Git) setPushStatus() {
```
I would argue we need to add a separate property to enable fetching this. Otherwise everyone will wait for this to complete. We can add `fetch_push_status` to avoid calling this.

sanki92 replied:
Implemented! Added `fetch_push_status` property... The `setPushStatus()` now only runs when both `fetch_status` and `fetch_push_status` are enabled

JanDeDobbeleer commented on src/segments/git.go:
```diff
@@ -538,6 +541,59 @@ func (g *Git) setBranchStatus() {
+		return ""
+	}
+
+	cfg, err := ini.Load(filepath.Join(g.scmDir, "config"))
```
can we check if we already loaded this? Because if so, we should have a separate function that caches the `cfg` instead.

sanki92 replied:
Fixed! ..implemented `getGitConfig()` that checks `g.gitConfig` first and only loads from disk if nil. Caches for future calls

JanDeDobbeleer reviewed (changes requested):
Not bad at all, like this. Few pointers to make it more efficient.

sanki92 committed:
refactor(git): add fetch_push_status property and cache git config

- Add fetch_push_status property to gate push-remote fetching
- Implement getGitConfig() to cache .git/config for performance
- Update documentation for the new property

Addresses maintainer feedback on PR #6851

sanki92 committed:
refactor(git): use strings.Index instead of regex for upstream parsing

Replaced regex pattern matching with simpler strings.Index approach for better performance and readability.

sanki92 commented on src/segments/git.go:
```diff
@@ -538,6 +546,73 @@ func (g *Git) setBranchStatus() {
+}
+
+func (g *Git) getPushRemote() string {
+	upstream := regex.ReplaceAllString("/.*", g.Upstream, "")
```
Good catch! Replaced regex with `strings.Index` for better performance and simplicity. Much cleaner approach.

sanki92 commented on website/docs/segments/scm/git.mdx:
```diff
@@ -53,6 +53,7 @@ You can set the following properties to `true` to enable fetching additional inf
 | Name                  |        Type         | Default | Description                                                                                                                                                                                                                                                                                                                           |
 | --------------------- | :-----------------: | :-----: | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | `fetch_status`        |      `boolean`      | `false` | fetch the local changes                                                                                                                                                                                                                                                                                                               |
+| `fetch_push_status`   |      `boolean`      | `false` | fetch the push-remote ahead/behind information. Requires `fetch_status` to be enabled                                                                                                                                                                                                                                                 |
```
Actually, the requirement is enforced through nesting. `setPushStatus()` on line 214 is inside the `if displayStatus` block that starts on line 209, so `fetch_status` must be `true` for `setPushStatus()` to execute.

JanDeDobbeleer committed:
refactor(git): only fetch git config once

JanDeDobbeleer committed:
refactor(git): move enabled check inside function

JanDeDobbeleer committed:
chore(git): add missing coverage test

JanDeDobbeleer committed:
chore(git): align struct fields

JanDeDobbeleer commented:
@all-contributors please add @sanki92 for code,doc

This is pure magic 🪄! @holopin-bot @sanki92 wizard
